src/lib.rs

```diff
@@ -40,19 +42,314 @@ mod repl_helper;
+        /// configuration with unpredictable results.  
+        ///
+        /// # Arguments
+        ///
+        /// - `machine`: A mutable pointer to the [`Machine`] instance.
+        /// - `query_state`: A mutable pointer to the [`QueryState`] instance.
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_cleanup_query_generator(machine: *mut Machine, query_state: *mut QueryState) {
+            unsafe {
+                if !query_state.is_null() {
+                    drop(Box::from_raw(query_state));
+                }
+                (*machine).cleanup_query_generator();
+            }
+        }
+
+        /// To be used to power iterator workflows, return a single goal of a Prolog query.
+        ///
+        /// # Safety
+        ///
+        /// * `machine` must be a valid [`Machine`] pointer created with [`scryer_machine_new`] 
+        /// that has not yet been freed.
+        /// * `query_state` must be a valid [`QueryState`] created with
+        /// [`scryer_start_new_query_generator`] that has not yet been freed.
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by 
+        /// [`scryer_start_new_query_generator`] that has not yet been freed with 
+        /// [`scryer_cleanup_query_generator`] or the [`Machine`] state will enter an undefined 
+        /// configuration with unpredictable results.  
+        /// * it is the responsibility of the caller to deallocate the pointer returned by 
+        /// this function with [`scyer_free_c_string`] in order to avoid memory leaks.
+        ///
+        /// # Returns
+        /// - Returns a pointer to a JSON formatted, null terminated UTF-8 string
+        ///
+        /// # Expected Response Format
+        /// ```json
+        /// // if result is a binding
+        ///
+        /// // current limitation is that only concrete (equality) bindings are returned,
+        /// // residual goals not yet supported.
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": [{ ... }], // single map entry, otherwise same signature as `scryer_run_query()`
+        /// }
+        ///
+        /// // if result is a boolean goal
+        ///
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": boolean // note -- `scryer_run_query_generator()` will continue returning results if
+        ///                     //       .. successively invoked, but the result will always be the same
+        ///                     //       .. at this point, `scryer_cleanup_query_generator()` should be called 
+        /// }        
+        ///
+        /// // if panic
+        /// { 
+        ///   "status": "error" | "panic",
+        ///   "error": error message | "panic"
+        /// }
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_run_query_generator(machine: &mut Machine, query_state: *mut QueryState) -> *mut c_char {
```

Comment:
Also, I think you can do the same thing you did with `Machine` and use a `&mut` for `QueryState` too. This would make it automatically a valid pointer to a `QueryState`, which would be automatically upheld in the Rust side. This _still_ needs to be `unsafe` as it is now because there is still the contract that the `QueryState` needs to have come from this `Machine`.

After #2472, this function should take just `&mut QueryState` because it will mutably borrow it's `Machine`, and because of this the contract that I mentioned will no longer be relevant and this function can drop the `unsafe`.

```suggestion
        pub unsafe extern "C" fn scryer_query_state_next(machine: &mut Machine, query_state: &mut QueryState) -> *mut c_char {
```